bpf-gpl/tc.c

```diff
@@ -728,16 +730,87 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct __sk_buff *skb,
 	}
 
 	l3_csum_off = skb_iphdr_offset(skb) +  offsetof(struct iphdr, check);
-	switch (state->ip_proto) {
-	case IPPROTO_TCP:
-		l4_csum_off = skb_l4hdr_offset(skb, ihl) + offsetof(struct tcphdr, check);
-		break;
-	case IPPROTO_UDP:
-		l4_csum_off = skb_l4hdr_offset(skb, ihl) + offsetof(struct udphdr, check);
-		break;
+
+	if (ct_related) {
+		if (ip_header->protocol == IPPROTO_ICMP) {
+			struct icmphdr *icmp;
+			bool outer_ip_snat;
+
+			/* if we do SNAT ... */
+			outer_ip_snat = ct_rc == CALI_CT_ESTABLISHED_SNAT;
+			/* ... there is a return path to the tunnel ... */
+			outer_ip_snat = outer_ip_snat && state->ct_result.tun_ret_ip;
+			/* ... and should do encap and it is not DSR or it is
+			 * it is leaving host and either DSR from WEP or
+			 * originated at host ... */
+			outer_ip_snat = outer_ip_snat &&
+				((dnat_return_should_encap() && !CALI_F_DSR) ||
+				 (CALI_F_TO_HEP &&
+				  ((CALI_F_DSR && skb_seen(skb)) || !skb_seen(skb))));
+
+			/* ... then fix the outer header IP first */
+			if (outer_ip_snat) {
+				ip_header->saddr = state->ct_result.nat_ip;
+				int res = bpf_l3_csum_replace(skb, l3_csum_off,
+						state->ip_src, state->ct_result.nat_ip, 4);
+				if (res) {
+					reason = CALI_REASON_CSUM_FAIL;
+					goto deny;
+				}
+				CALI_DEBUG("ICMP related: outer IP SNAT to %x\n",
+						be32_to_host(state->ct_result.nat_ip));
+			}
+
+			if (!icmp_skb_get_hdr(skb, &icmp)) {
+				CALI_DEBUG("Ooops, we already passed one such a check!!!\n");
+				goto deny;
+			}
+
+			l3_csum_off += sizeof(*ip_header) + sizeof(*icmp);
+			ip_header = (struct iphdr *)(icmp + 1); /* skip to inner ip */
+
+			/* flip the direction, we need to reverse the original packet */
+			switch (ct_rc) {
+			case CALI_CT_ESTABLISHED_SNAT:
+				/* handle the DSR case, see CALI_CT_ESTABLISHED_SNAT where nat is done */
+				if (dnat_return_should_encap() && state->ct_result.tun_ret_ip) {
+					if (CALI_F_DSR) {
+						/* SNAT will be done after routing, when leaving HEP */
+						CALI_DEBUG("DSR enabled, skipping SNAT + encap\n");
+						goto allow;
+					}
+				}
+				ct_rc = CALI_CT_ESTABLISHED_DNAT;
+				break;
+			case CALI_CT_ESTABLISHED_DNAT:
+				ct_rc = CALI_CT_ESTABLISHED_SNAT;
+				break;
+			}
+		}
 	}
 
-	switch (state->ct_result.rc){
+	struct tcphdr *tcp_header = (void*)(ip_header+1);
+	struct udphdr *udp_header = (void*)(ip_header+1);
+
+	__u8 ihl = ip_header->ihl * 4;
+
+	int res = 0;
+	bool encap_needed = false;
+
+	if (state->ip_proto == IPPROTO_ICMP && ct_related) {
+		/* do not fix up embedded L4 checksum for related ICMP */
+	} else {
+		switch (ip_header->protocol) {
```

Comment:
`state` carries information about the entire packet, the `state->ip_proto` is still icmp, but we have moved the `ip_header` to point to the inner header. All the editing it done where point to, so we should use that info even in the case of regular NAT